devito/types/dense.py

```diff
@@ -1360,6 +1361,31 @@ def __shape_setup__(cls, **kwargs):
                 shape.insert(cls._time_position, save)
             else:
                 raise TypeError("`save` can be None, int or Buffer, not %s" % type(save))
+        elif dimensions is None:
+            raise TypeError("`dimensions` required if both `grid` and "
+                            "`shape` are provided")
+        else:
+            # Got `grid`, `dimensions`, and `shape`. We sanity-check that the
+            # Dimensions in `dimensions` also appearing in `grid` have same size
+            # (given by `shape`) as that provided in `grid`
+            if len(shape) != len(dimensions):
+                raise ValueError("`shape` and `dimensions` must have the "
+                                 "same number of entries")
+            loc_shape = []
+            for d, s in zip(dimensions, shape):
+                if d in grid.dimensions:
+                    size = grid.dimension_map[d]
+                    if size.glb != s and s is not None:
```

Comment:
can `s` ever be None?

Reply:
This is just lifted from the `Function` implementation. But I think the idea is that you shouldn't have to specify the shape component for dimensions that are part of the Grid.

So you should be able to `TimeFunction(name='u', grid=grid, dimensions=(grid.time_dim, grid.dimensions[0]), shape=(10, None))`

Reply:
Ok we'll probably go over it another time then,